### DISCUSSION TITLE
`cuda::std::complex<float>` & `cufftComplex`

### DISCUSSION THREAD

**Discussion by ax3l** [NVIDIA/cccl#2247] [Q&A]
Hi there,

Are there any existing or planned formal guarantees that one could use a `cuda::std::complex<float>` (or double) for cuFFT operations?
In particular, can on just `reinterpret_cast` pointers to each others arrays?

I could not find any hints on  libc++ / cccl type compatibility in the [cuFFT docs](https://docs.nvidia.com/cuda/cufft/index.html#cufftcomplex).

From the headers, it looks like cuFFT complex types are based on `float2`/`double2`, which should have the same real+imag order [but different alignment](https://github.com/NVIDIA/libcudacxx/issues/151)? Sooo.... not safe to `reinterpret_cast`?

**Comment by miscco** [comment]
We do not provide any guarantees about binary compatibility with types from other libraries. That would tie both our hands and potentially prevent us from making any changes to the types.

I would recommend to verify that alignment and size of the two types is the same and then move from that on your own risk.

Also note that we have slightly different requirements than cuFFT especially when it comes to constexpr evaluation so it is not as trivial to internally use the float2 / double2 vector types. (I tried ...)

**Reply by ax3l** [reply]
That makes sense. I think this is more a cuFFT feature request to provide additional APIs than a cccl RFE.

**Comment by fbusato** [comment]
@ax3l this is not only relevant  for cuFFT. All CUDA Math libs (cuBLAS, cuSPARSE, cuSOLVER, etc.) use `cuComplex` and `cuDoubleComplex` (`float2`/`double2`). Conversion between C and C++ types is an extremely common situation and we should provide some kind of guarantees.

**Reply by fbusato** [reply]
@ax3l btw, `reinterpret_cast` is undefined behavior. You should prefer `cuda::std::bit_cast`

**Comment by fbusato** [comment]
@miscco I was not aware that `std::complex` already provides this guarantee https://en.cppreference.com/w/cpp/numeric/complex [**Array-oriented access**] (Jake suggestion)

**Reply by miscco** [reply]
Oh then we should support that definitely

**Reply by ax3l** [reply]
@fbusato @miscco any news on this end? :)

**Reply by miscco** [reply]
Oh as I said we are conforming, so reinterpret_cast should just work. 

We are explicitly aligning to 2* sizeof(T) so that should be fine
